### PR TITLE
10 add p block

### PR DESCRIPTION
**Pull request by Bartlomiej21**
Adding PBlock class and tests for it

### CODE REVIEW

**Comment by marcin-ogorzalek-EPAM** [comment]
No lined/closed issue int the description.

File: src/main/java/com/epam/prejap/teatrees/block/PBlock.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package com.epam.prejap.teatrees.block;
+/**
+ *
+ * This class introduces the P-shaped block in Tetris game.
+ *
+ * @author Bartłomiej Słomiński
+ *
+ */
+
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/block/PBlock.java:9]: An unnecessary empty line. Javadoc should be directly above the thing it is describing.

File: src/main/java/com/epam/prejap/teatrees/block/PBlock.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package com.epam.prejap.teatrees.block;
+/**
+ *
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/block/PBlock.java:3]: There is no need for an empty line at the beginning of javadoc.

File: src/main/java/com/epam/prejap/teatrees/block/PBlock.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * This class introduces the P-shaped block in Tetris game.
+ *
+ * @author Bartłomiej Słomiński
+ *
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/block/PBlock.java:7]: There is no need for an empty line at the beginning of javadoc.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Test
+    void checkIfBlocksThatAreNotRectangularsWillPass(){
+        //every line should have the same length
+        int rowLength = WRONG_IMAGE[0].length;
+        for (int i=0; i<WRONG_IMAGE.length; i++){
+            assertEquals(WRONG_IMAGE[i].length,rowLength);
+        }
+    }
+}
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:89]: No empty line at the end of the file. This is a good practice to leave one (some tools or scripting may require this). Configure your IDE so it will be added automatically.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        for (int i=0; i<IMAGE.length; i++){
+            assertEquals(IMAGE[i].length,rowLength);
+        }
+    }
+
+    //NOT SO HAPPY PATH
+    @Test
+    void checkIfVoidRowsAreDetected(){
+        for (int i=0; i<WRONG_IMAGE.length;i++){
+            assertTrue(WRONG_IMAGE[i].length>0);
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:59]: Why not get rid of `WRONG_IMAGE` and instead use `assert false`?

Applies here and onwards.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                assertTrue(field>=0 && field <=1 );
+            }
+        }
+    }
+
+    @Test
+    void checkIfIncorrectRowLengthIsDetected(){
+        int columns = WRONG_IMAGE[0].length;
+        for (int i=0; i<WRONG_IMAGE.length; i++){
+            assertEquals(WRONG_IMAGE[i].length,columns);
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:77]: Same thing as previously bu `assertNotEquals`.

Hint: use `IMAGE`

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+
+    @Test
+    void checkIfBlocksThatAreNotRectangularsWillPass(){
+        //every line should have the same length
+        int rowLength = WRONG_IMAGE[0].length;
+        for (int i=0; i<WRONG_IMAGE.length; i++){
+            assertEquals(WRONG_IMAGE[i].length,rowLength);
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:86]: The previous remark applies.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            {1, 5},
+            {1, 0},
+            {},
+    };
+
+    //HAPPY PATH
+    @Test
+    void checkIfNoRowsAreVoid(){
+        for (int i=0; i<IMAGE.length;i++){
+            assertTrue(IMAGE[i].length>0);
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:24]: If you have more than one assertion in a test (those in fors applies), then you should use soft assertion form TestNG: https://www.softwaretestingmaterial.com/soft-assert/

Comment (marcin-ogorzalek-EPAM) [reply src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:24]: Applies here and onwards

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+
+    @Test
+    void checkIfBlockFieldsContainOnlyOnesAndZeroes(){
+        for (int i=0; i<IMAGE.length; i++){
+            for (int j=0; j<IMAGE[0].length;j++) {
+                byte field = IMAGE[i][j];
+                assertTrue(field>=0 && field <=1 );
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:33]: I would divide this into two separate `assertTrue`

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    final byte[][] WRONG_IMAGE = {
+            {1, 1, 1},
+            {1, 5},
+            {1, 0},
+            {},
+    };
+
+    //HAPPY PATH
+    @Test
+    void checkIfNoRowsAreVoid(){
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:22]: BTW, you know that you are testing here not the instances of `PBlock()` but hardcoded matrixes (`IMAGE` and `WRONG_IMAGE`). This doesn't make sense.

File: src/main/java/com/epam/prejap/teatrees/block/PBlock.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+
+    public PBlock() {
+        super(IMAGE);
+    }
+
+}
+
+
+
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/block/PBlock.java:26]: Too many empty lines at the end. One is perfectly enough.

**Comment by marcin-ogorzalek-EPAM** [comment]
Also, you have to bump up (increase) the project's version in the `pom.xml`.

**Comment by Bartlomiej21** [comment]
I have finished my changes to the code and I'm waiting for a review.

**Comment by marcin-ogorzalek-EPAM** [comment]
Link this pull request with a proper issue by comment `Closes #10` in the description

File: src/main/java/com/epam/prejap/teatrees/block/PBlock.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
+package com.epam.prejap.teatrees.block;
+/**
+ * This class introduces the P-shaped block in a Tetris game.
+ * @author Bartłomiej Słomiński
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/block/PBlock.java:4]:
```suggestion
 * This class introduces the P-shaped block in a Tetris game.
 *
 * @author Bartłomiej Słomiński
```

Yet `@author` should demlimeted from the class overview

**Comment by marcin-ogorzalek-EPAM** [comment]
Update project's version in `pom.xml` to at least `0.8` at this point

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            {1, 1},
+            {1, 0},
+            {1,}
+    };
+    final byte[][] WRONG_IMAGE3 = {
+            {1, 1},
+            {1, -1},
+            {1, 0},
+            {0, 1},
+    };
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:20]: Those could be enclosed in the data provider. That would give you the possibility to get rid of repetitive tests and keep code DRY.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+    //NOT SO HAPPY PATH
+    @Test (expectedExceptions = {IllegalArgumentException.class}, expectedExceptionsMessageRegExp = "Image has height equal to 0")
+    void checkIfBlockWithZeroHeightIsCaught() {
+        try {
+            new PBlock(WRONG_IMAGE);
+        }
+        catch (Exception e){
+            throw e;
+        }
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:66]: Don't catch and rethrow. Use `throws` in method declaration instead

Comment (marcin-ogorzalek-EPAM) [reply src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:66]: And do that with a specified exception, not the most general one.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            {1, 0},
+            {1,}
+    };
+    final byte[][] WRONG_IMAGE3 = {
+            {1, 1},
+            {1, -1},
+            {1, 0},
+            {0, 1},
+    };
+    PBlock block = new PBlock();
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:21]: this field should be initialised in `private void setup()` method annotated with `@BeforeMethod`. Moreover, the filed itself should be `private`

Comment (marcin-ogorzalek-EPAM) [reply src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:21]: Right now this doesn't give us a guarantee that tests are independent.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
+package com.epam.prejap.teatrees.block;
+
+import org.testng.annotations.Test;
+import org.testng.asserts.SoftAssert;
+
+public class PBlockTest {
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:6]: Add those test to proper grope (via annotation and in `testng.xml`

**Comment by marcin-ogorzalek-EPAM** [comment]
Update your fork with what is on `master` in the main repo.

Update project's version in `pom.xml`. Update `testng.xml`. Resolve merge conflicts

**Comment by Bartlomiej21** [comment]
Issue 10

Closes #10

What was changed:
Added PBlock class
Added tests for PBlock

**Review by marcin-ogorzalek-EPAM — CHANGES_REQUESTED** [review]
There are still things to impove

**Review by pkierat-epam — CHANGES_REQUESTED** [review]
The pom version is not updated.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+                return new Object[][]{
+                        {WRONG_IMAGE1},
+                };
+            case "checkIfWrongDotValueIsDetected":
+                return new Object[][]{
+                        {WRONG_IMAGE2},
+                };
+        }
+        return null;
+    }
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:37]: This doesn't look like the data provider that I'm used to. What happening here?

Comment (pkierat-epam) [reply src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:37]: @marcin-ogorzalek-EPAM I think I know: a single data provider that serves different data based on the name of the method it's applied to.

@Bartlomiej21 Although it's quite interesting (I didn't even know TestNG can do that), have you heard about SRP? Moreover, what will happen, if someone uses this DP for a test method which is not supported by the switch? Splitting that into 3 independent data provider methods would actually save you a lot of code.

Comment (Bartlomiej21) [reply src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java]: After deliberation, I have concluded that my tests shouldn't include things like testing if the image is empty, since it's more of a Block class test rather than PBlock. I have left it for now but maybe it should be moved?
In my most recent push I've made some changes to DataProviders and hope it'll be OK.

File: pom.xml [hunk 18]
```diff
@@ -6,7 +6,7 @@
 
     <groupId>com.epam.prejap</groupId>
     <artifactId>tea-trees</artifactId>
-    <version>0.11</version>
+    <version>0.12</version>
```
Comment (marcin-ogorzalek-EPAM) [inline pom.xml:9]: The current version is 0.15

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+        softAssertion.assertEquals(block.dotAt(row,col),dot);
+        softAssertion.assertAll();
+    }
+
+    @Test (expectedExceptions = {IllegalArgumentException.class}, expectedExceptionsMessageRegExp = "Image has height equal to 0")
+    void checkIfBlockWithZeroHeightIsCaught() throws IllegalArgumentException{
+        new PBlock(EMPTY_IMAGE);
+    }
+
+    @Test (dataProvider = "NonRectangularImage", expectedExceptions = {IllegalArgumentException.class}, expectedExceptionsMessageRegExp = "Image is not a rectangle")
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:99]: Line lengthier than 120 characters.  Shouldn't be. Move ` expectedExceptionsMessageRegExp` and all other rest to the next line.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+        softAssertion.assertTrue(onePresent);
+        softAssertion.assertAll();
+    }
+
+    @Test
+    void checkIfNumberOfColumnsIsCorrect(){
+        SoftAssert softAssert = new SoftAssert();
+        softAssert.assertEquals(block.cols(),2);
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:75]: Don't use soft assertion when you have only one `assertXXX` in the test

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+            softAssertion.assertEquals(block.image[i].length,block.cols);
+        }
+        softAssertion.assertAll();
+    }
+
+    @Test (dataProvider = "PCoordinates")
+    void checkIfCoordinatesForPBlockAreCorrect(int row, int col, int dot){
+        SoftAssert softAssertion = new SoftAssert();
+        softAssertion.assertEquals(block.dotAt(row,col),dot);
+        softAssertion.assertAll();
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:91]: Don't use soft assertion when you have only one assertXXX in the test

**Comment by asilenko** [comment]
PBlock must be included in com.epam.prejap.teatrees.block.BlockFeed
Maybe the list could get a vertical structure and alphabetical order by the way?

File: pom.xml [hunk 22]
```diff
@@ -6,7 +6,7 @@
 
     <groupId>com.epam.prejap</groupId>
     <artifactId>tea-trees</artifactId>
-    <version>0.11</version>
+    <version>0.16</version>
```
Comment (marcin-ogorzalek-EPAM) [inline pom.xml:9]: Please sync the branch with main repo

**Comment by Bartlomiej21** [comment]
> PBlock must be included in com.epam.prejap.teatrees.block.BlockFeed Maybe the list could get a vertical structure and alphabetical order by the way?
I have added PBlock to BlockFeed. Don't know about the vertical list, but alphabetical order would be good.

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+                if (field==1) onePresent=true;
+            }
+        }
+        softAssertion.assertTrue(onePresent);
+        softAssertion.assertAll();
+    }
+
+    @Test
+    void checkIfNumberOfColumnsIsCorrect(){
+        Assert.assertEquals(block.cols(),2);
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:75]: Static import for assertion could be used here (and onwards), but I will let it slip. Minor issue

File: src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+    final byte[][] EMPTY_IMAGE = {};
+
+    @BeforeMethod
+    void setup() {
+        block = new PBlock();
+    }
+
+    @Test
+    void checkIfNoRowsAreVoid(){
+        SoftAssert softAssertion= new SoftAssert();
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:50]: Slight issue with formatting (missing space). Minor issue

Comment (marcin-ogorzalek-EPAM) [reply src/test/java/com/epam/prejap/teatrees/block/PBlockTest.java:50]: And not only here. I will let it slip.

**Review by marcin-ogorzalek-EPAM — CHANGES_REQUESTED** [review]
General remark. Check Tests for other blocks. Try to adjust the naming convention of your test, so those align. Additionally, Think about assertions with messages (as in those other tests). This is a good practice to add such information as what had failed in a given assertion.

Right now this Is good enough to be merged. Yet I will give you an hour or two to improve if you have time. 

I like that you added some additional tests (different from other blocks). Yet I thinking if those not checking super class `Block`  itself (exception testing).

**Comment by marcin-ogorzalek-EPAM** [comment]
Just let me know if you will be working on quick fixes or not @Bartlomiej21

**Comment by Bartlomiej21** [comment]
> Just let me know if you will be working on quick fixes or not @Bartlomiej21

Sorry, I had to take additional shift for a missing colleague in my work place, so I'm working two consecutive night shifts (yesterday and today) with little time between just to sleep and eat. Next time I'll have opportunity to do some code is tommorow morning. I can make some changes then.
About super class Block, I have thought the same - some tests are testing general block construction, not the PBlock itself.

**Comment by Bartlomiej21** [comment]
I have deleted some of the tests that should belong to Block super class (like testing exceptions). If there will be need, I can rewrite them as part of BlockTest.
